bin/transform

```diff
@@ -247,6 +251,95 @@ def update_metadata(curated_gisaid_data: pd.DataFrame) -> pd.DataFrame:
+    ``AssertionError`` if the tests fail.
+    """
+    def annotation_hierarchies(annotations: pd.DataFrame) -> Optional[pd.DataFrame]:
+        """
+        Returns a ``pd.DataFrame` of location hierarchies from the annotations
+        file if annotations are provided, else returns ``None``.
+        """
+        if not args.annotations:
+            return
+
+        annotations.columns = ['strain', 'gisaid_epi_isl', 'variable', 'value']
+        annotations = annotations \
+            .drop_duplicates(['gisaid_epi_isl', 'variable'], keep='last')
+
+        keep_columns = location_hierarchy_columns + \
+            [ f'{resolution}_exposure' for resolution in location_hierarchy_columns]
+        # Avoid a KeyError if the annotations file doesn't contain all of
+        # our resolution columns
+        columns = [ col for col in annotations.columns if col in keep_columns ]
+
+        wide_annotations = annotations \
+            .pivot(index='gisaid_epi_isl', columns='variable', values='value')
+        wide_annotations = wide_annotations[columns].dropna()
+
+        wide_annotations['gisaid_epi_isl'] = wide_annotations.index
+        wide_annotations = wide_annotations.reset_index(drop=True)
+
+        return hierarchy_dataframe(wide_annotations, 'gisaid_epi_isl',
+            location_hierarchy_columns)
+
+    def no_missing_hierarchy_data():
+        """
+        Raises an ``AssertionError`` if the metadata have missing values in the
+        location hierarchy columns.
+        """
+        assert metadata[location_hierarchy_columns].notnull().all().all(), \
+            "There are missing values in the location hierarchy columns " \
+            f"«{location_hierarchy_columns}». At this point, we expect all " \
+            "missing values to be interpolated."
+
+    def no_unknown_hierarchies(known_location_hierarchies: pd.DataFrame):
+        """
+        Produces a new ``pd.DataFrame`` of all new location hierarchies in the
+        curated metadata that don't exist in the given
+        *known_location_hierarchies* `pd.DataFrame``. If the resulting DataFrame
+        is not empty, prints it as a TSV to the specified output location.
+        """
+        metadata_hierarchies = hierarchy_dataframe(metadata, 'gisaid_epi_isl',
+            location_hierarchy_columns)
+
+        all_hierarchies = metadata_hierarchies.merge(
+            known_location_hierarchies,
+            how='outer',
+            indicator=True
+        )
+
+        unknown_hierarchies = all_hierarchies[all_hierarchies._merge == 'left_only']
+
+        if not unknown_hierarchies.empty:
+            unknown_hierarchies[location_hierarchy_columns] \
+                .to_csv(args.output_new_location_hierarchies, sep='\t', index=False)
+
+    # Combine metadata location hierarchies with those coded in annotations
+    known_location_hierarchies = pd.concat([
+        pd.read_csv(args.location_hierarchy, sep='\t'),
+        annotation_hierarchies(annotations)]
```

Comment:
I don't understand why this combines the location hierarchy file with the annotations file in order to produce known locations.

Separately, reading the annotations file as a table is not quite kosher and might lead to subtly incorrect results. This is because earlier lines in the annotations file can be overridden by later lines. It is essentially a procedural list of changes to make. To make it function more like an overlay table, we'd want to enforce uniqueness on (record id, field id).

Reply:
I think I was thinking in too-constrained terms when implementing the location hierarchy validation. The desire to also check the annotations file comes from the check I wanted to replace in _./bin/check-annotations_ that every country annotation also has a division annotation to match, etc.